daemon/image.go

```diff
@@ -113,12 +121,233 @@ func (daemon *Daemon) GetCachedImage(imgID image.ID, config *containertypes.Conf
+	for _, cacheFromImageName := range cacheFrom {
+		cacheFromImage, err := daemon.GetImage(cacheFromImageName)
+		if err != nil {
+			logrus.Warnf("Could not look up %s for cache resolution, skipping: %s", cacheFromImageName, err)
+			continue
+		}
+
+		cache.cacheFromImages[cacheFromImageName] = cacheFromImage
+		cache.cacheFromImageHistories[cacheFromImageName] = makeHistoryWithSource(cacheFromImage)
+	}
+
+	return cache
+}
+
+// In the history array, we have pairs of (command, resultingLayerID). What we actually want to be able
+// to compare is pairs of (sourceLayerID, command), and if we have a match, consult resultingLayerID.
+// We also don't directly have source/resultingLayerID, but rather a boolean "did create new layer" flag.
+// Define a struct to store this mapping for convenience.
+type historyWithSourceT struct {
+	// sourceLayerID is the layer on which the command was run. Empty digest if this is the first command or
+	// if nothing has actually been added to the rootfs yet.
+	sourceLayerID layer.DiffID
+	// cmd is the command which got run on sourceLayerID
+	cmd string
+	// resulingLayerID is the result of running cmd on sourceLayerID (might be the same as sourceLayerID)
+	resultingLayerID layer.DiffID
+	// createdAt is the time the history entry was created
+	createdAt time.Time
+	// history is the actual, underlying history entry
+	history image.History
+}
+
+func makeHistoryWithSource(image *image.Image) []historyWithSourceT {
+	// Let's make those structs now
+	historyWithSource := make([]historyWithSourceT, len(image.History))
+	layerIndex := -1
+	for i, h := range image.History {
+
+		// previous is layerIndex from previous iteration
+		if layerIndex == -1 {
+			historyWithSource[i].sourceLayerID = digest.DigestSha256EmptyTar
+		} else {
+			historyWithSource[i].sourceLayerID = image.RootFS.DiffIDs[layerIndex]
+		}
+
+		// now increment, if needed, and look at the result layer ID
+		if !h.EmptyLayer {
+			layerIndex = layerIndex + 1
+		}
+		if layerIndex == -1 {
+			historyWithSource[i].resultingLayerID = digest.DigestSha256EmptyTar
+		} else {
+			historyWithSource[i].resultingLayerID = image.RootFS.DiffIDs[layerIndex]
+		}
+
+		// Copy the other history entries over I'm interested in
+		historyWithSource[i].cmd = h.CreatedBy
+		historyWithSource[i].createdAt = h.Created
+		historyWithSource[i].history = h
+	}
+
+	return historyWithSource
+}
+
 // GetCachedImageOnBuild returns a reference to a cached image whose parent equals `parent`
 // and runconfig equals `cfg`. A cache miss is expected to return an empty ID and a nil error.
-func (daemon *Daemon) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
-	cache, err := daemon.GetCachedImage(image.ID(imgID), cfg)
-	if cache == nil || err != nil {
+func (cache *daemonImageCacheForBuild) GetCachedImageOnBuild(imgID string, cfg *containertypes.Config) (string, error) {
+	cachedImage, err := cache.daemon.GetCachedImage(image.ID(imgID), cfg)
```

Comment:
Once `cache-from` is used we should only match these images. Probably just need a check that this image is in the `cache-from` parent chain.